src/handle_action.cpp

```diff
@@ -769,15 +770,18 @@ static void sleep()
             as_m.entries.emplace_back( uimenu_entry( i, true, '0' + i,
                                        string_format( _( "Set alarm to wake up in %i hours." ), i ) ) );
         }
-    }
 
-    as_m.query();
-    if( as_m.ret >= 3 && as_m.ret <= 9 ) {
-        u.add_effect( effect_alarm_clock, 1_hours * as_m.ret );
+        as_m.query();
+        if( as_m.ret >= 3 && as_m.ret <= 9 ) {
+            u.add_effect( effect_alarm_clock, 1_hours * as_m.ret );
+            try_sleep_dur = 1_hours * as_m.ret;
```

Comment:
I just noticed this line sets the `try_sleep_dur` to the time-to-alarm; that explains why one time my player gave up trying to sleep before the alarm went off while the second time I got the expected beep-beep-beep and such. At a guess, sometimes the activity will process before the `effect_alarm_clock`, and other times the alarm will process first.

I don't think that's unrealistic (I imagine the character taking a look at the clock, seeing the alarm is about to trigger, and just abandoning the sleep attempt -- something I've done myself at times), just mentioning something you might've missed.